lib/gringotts/gateways/sagepay.ex

```diff
@@ -282,62 +305,79 @@ defmodule Gringotts.Gateways.Sagepay do
-      |> Map.get("merchantSessionKey")
+    {:ok, merchant_key} = commit(:post, "merchant-session-keys", vender_name, merchant_header)
+
+    merchant_key
+    |> Map.get("merchantSessionKey")
+  end
+
+  defp card_params(card) do
+    %{
+      "cardDetails" => %{
+        "cardholderName" => CreditCard.full_name(card),
+        "cardNumber" => card.number,
+        "expiryDate" =>
+          (card.month * 100 + card.year)
+          |> Integer.to_string()
+          |> String.pad_leading(4, "0"),
+        "securityCode" => card.verification_code
+      }
+    }
   end
 
   # Function generate_ckey generate a unique cardidentifier for every transaction
 
-  defp generate_ckey(card, merchant_key) do
+  defp generate_card_identifier(card, merchant_key) do
     card_header = [
       {"Authorization", "Bearer " <> merchant_key},
       {"Content-type", "application/json"}
     ]
 
     card = card |> Poison.encode!()
 
-    card_identifier =
-      commit(:post, "card-identifiers", card, card_header)
-      |> Map.get("cardIdentifier")
+    {:ok, card_identifier} = commit(:post, "card-identifiers", card, card_header)
+
+    card_identifier
+    |> Map.get("cardIdentifier")
   end
 
-  # Function card_details creates the actual body (details of the customer )of the card 
+  # Function transaction_details creates the actual body (details of the customer )of the card 
   # and with merchant_session_key, card_identifiier ,shipping address of a
   # customer, and other details and converting the map into keyword list
 
-  defp card_details(amount, merchant_key, card_key, opts) do
-    card_body =
-      Poison.encode!(%{
-        "transactionType" => opts[:transactionType],
-        "paymentMethod" => %{
-          "card" => %{
-            "merchantSessionKey" => merchant_key,
-            "cardIdentifier" => card_key,
-            "save" => false
-          }
-        },
-        "vendorTxCode" => opts[:vendorTxCode],
-        "amount" => amount,
-        "currency" => opts[:currency],
-        "description" => opts[:description],
-        "customerFirstName" => opts[:customerFirstName],
-        "customerLastName" => opts[:customerLastName],
-        "billingAddress" => %{
-          "address1" => opts[:billingAddress].address1,
-          "city" => opts[:billingAddress].city,
-          "postalCode" => opts[:billingAddress].postalCode,
-          "country" => opts[:billingAddress].country
+  defp transaction_details(amount, merchant_key, card_identifiier, opts) do
+    {currency, value} = Money.to_string(amount)
+
+    Poison.encode!(%{
+      "transactionType" => opts[:transactionType],
+      "paymentMethod" => %{
+        "card" => %{
+          "merchantSessionKey" => merchant_key,
+          "cardIdentifier" => card_identifiier,
+          "save" => false
         }
-      })
+      },
+      "vendorTxCode" => opts[:vendorTxCode],
+      "amount" => Kernel.trunc(String.to_float(value)),
```

Comment:
You perhaps want to use `Money.to_integer/1`. It'll be easier to figure this out if you look at [`activemerchant`](https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/sage_pay.rb) implementation of SagePay.